ping/ping6_common.c

```diff
@@ -571,6 +577,90 @@ int build_echo(struct ping_rts *rts, uint8_t *_icmph,
 	return cc;
 }
 
+int build_probe(struct ping_rts *rts, uint8_t *_icmph,
+	       unsigned packet_size __attribute__((__unused__)))
+{
+	struct exthdr *extbase, ext;
+	struct iiohdr *iiobase, iio;
+	struct icmp6_hdr *icmph;
+	uint32_t iio_ip_hdr = 0;
+	int cc;
+
+	icmph = (struct icmp6_hdr *)_icmph;
+	extbase = (struct exthdr *)(icmph + 1);
+	iiobase = (struct iiohdr *)((char *)extbase + sizeof(struct exthdr));
+	icmph->icmp6_type = ICMPV6_EXT_ECHO_REQUEST;
+	icmph->icmp6_code = 0;
+	icmph->icmp6_cksum = 0;
+	icmph->icmp6_id = rts->ident;
+	/* PROBE messages use only the first 8 bits as sequence number */
+	icmph->icmp6_dataun.icmp6_un_data8[2] = rts->ntransmitted + 1;
+	icmph->icmp6_dataun.icmp6_un_data8[3] = 1;	/* Set L-bit */
```

Comment:
I recommend giving the user an argument to control this (e.g., "--remote") - this is one of the most important pieces of PROBE functionality.  In the case where the L-bit is clear, the only valid ctype is ICMP_EXT_ECHO_CTYPE_ADDR, but otherwise all the rest of the code applies.